tests/test_websockets.py

```diff
@@ -0,0 +1,242 @@
+import base64
+import re
+import hashlib
+from collections import OrderedDict
+
+from six.moves import socketserver
+import pytest
+import paho.mqtt.client as client
+
+from paho.mqtt.client import WebsocketConnectionError
+from testsupport.broker import fake_websocket_broker
+
+
+@pytest.fixture
+def init_response_headers():
+    # "Normal" websocket response from server
+    response_headers = OrderedDict([
+        ("Upgrade", "websocket"),
+        ("Connection", "Upgrade"),
+        ("Sec-WebSocket-Accept", "testwebsocketkey"),
+        ("Sec-WebSocket-Protocol", "chat"),
+    ])
+
+    return response_headers
+
+
+def get_websocket_response(response_headers):
+    """ Takes headers and constructs HTTP response
+
+    'HTTP/1.1 101 Switching Protocols' is the headers for the response,
+    as expected in client.py
+    """
+    response = "\r\n".join([
+        "HTTP/1.1 101 Switching Protocols",
+        "\r\n".join("{}: {}".format(i, j) for i, j in response_headers.items()),
+        "\r\n",
+    ]).encode("utf8")
+
+    return response
+
+
+@pytest.mark.parametrize("proto_ver,proto_name", [
+    (client.MQTTv31, "MQIsdp"),
+    (client.MQTTv311, "MQTT"),
+])
+class TestInvalidWebsocketResponse(object):
+    def test_unexpected_response(self, proto_ver, proto_name, fake_websocket_broker):
+        """ Server responds with a valid code, but it's not what the client expected """
+
+        mqttc = client.Client(
+            "test_unexpected_response",
+            protocol=proto_ver,
+            transport="websockets"
+            )
+
+        with fake_websocket_broker.serve("200 OK\n"):
```

Comment:
TCPServer don't expect a string. It cause the server to close the connection immediately.